src/Tools/dotnet-monitor/Commands/ConfigShowCommandHandler.cs

```diff
@@ -15,19 +16,42 @@ internal sealed class ConfigShowCommandHandler
         // Although the "noHttpEgress" parameter is unused, it keeps the entire command parameter set a superset
         // of the "collect" command so that users can take the same arguments from "collect" and use it on "config show"
         // to get the same configuration without telling them to drop specific command line arguments.
-        public static void Invoke(string[] urls, string[] metricUrls, bool metrics, string diagnosticPort, bool noAuth, bool tempApiKey, bool noHttpEgress, ConfigDisplayLevel level)
+        public static void Invoke(string[] urls, string[] metricUrls, bool metrics, string diagnosticPort, bool noAuth, bool tempApiKey, bool noHttpEgress, ConfigDisplayLevel level, bool showSources)
         {
-            Write(Console.OpenStandardOutput(), urls, metricUrls, metrics, diagnosticPort, noAuth, tempApiKey, level);
+            Write(Console.OpenStandardOutput(), urls, metricUrls, metrics, diagnosticPort, noAuth, tempApiKey, level, showSources);
         }
 
-        public static void Write(Stream stream, string[] urls, string[] metricUrls, bool metrics, string diagnosticPort, bool noAuth, bool tempApiKey, ConfigDisplayLevel level)
+        public static void Write(Stream stream, string[] urls, string[] metricUrls, bool metrics, string diagnosticPort, bool noAuth, bool tempApiKey, ConfigDisplayLevel level, bool showSources)
         {
             IAuthConfiguration authConfiguration = HostBuilderHelper.CreateAuthConfiguration(noAuth, tempApiKey);
             HostBuilderSettings settings = HostBuilderSettings.CreateMonitor(urls, metricUrls, metrics, diagnosticPort, authConfiguration);
             IHost host = HostBuilderHelper.CreateHostBuilder(settings).Build();
             IConfiguration configuration = host.Services.GetRequiredService<IConfiguration>();
             using ConfigurationJsonWriter jsonWriter = new ConfigurationJsonWriter(stream);
             jsonWriter.Write(configuration, full: level == ConfigDisplayLevel.Full, skipNotPresent: false);
+
+            if (showSources)
+            {
+                WriteConfigurationProviders(stream, configuration);
+            }
+        }
+
+        private static void WriteConfigurationProviders(Stream stream, IConfiguration configuration)
+        {
+            var configurationProviders = ((IConfigurationRoot)configuration).Providers.Reverse();
+
+            StreamWriter writer = new StreamWriter(stream);
+
+            writer.WriteLine("Configuration Providers (High to Low Priority):");
+
+            foreach (var provider in configurationProviders)
```

Comment:
While I think this is going to be very useful, how difficult would it be to do it on a per value basis? Consider something like IConfigurationRoot.GetDebugView() https://github.com/dotnet/runtime/blob/50e6f078bcf4aa5bca1e7e110b9a6cb846390b82/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationRootExtensions.cs#L36

Reply:
I agree with this sentiment. I'm hoping we can do something more like [GetDebugView](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.configurationrootextensions.getdebugview?view=dotnet-plat-ext-3.1). Here's example output of calling this: https://stackoverflow.com/questions/37688027/how-can-i-list-all-of-the-configuration-sources-or-properties-in-asp-net-core

I'm thinking that we could add the source as a comment above the JSON element using [Utf8JsonWriter.WriteCommentValue](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonwriter.writecommentvalue?view=net-6.0) so that it is still valid JSON.

We can also do better than what GetDebugView writes out by writing useful information for each provider type:
- EnvironmentVariablesConfigurationProvider would write out the prefix
- JsonConfigurationProvider would write out the file path or name
- KeyPerFileConfigurationProvider would write out the directory

etc

Reply:
+1 on the Json comments. This also removes the need for an additional non-json text block at the end of the output.

Reply:
Maybe just start with writing out the provider type and then do a follow up PR to write more specific information.

Reply:
To further clarify, I would like to see a comment added to every value as to where that value came from in configuration, not just running through the sources and printing them at the top (because everyone is going to have the same list), e.g.

```
{
  // Source: Environment variable with prefix 'DOTNET_'
  "urls": "http://localhost:8080"
  // Source: JSON file 'settings.json'
  "DiagnosticPort": "/diag/port"
}
```

Reply:
Ah, that makes sense - I'll work on that

Reply:
So, I've gotten a version of this working...my only concern is that readability isn't great. Do people have thoughts about having the sources listed at the top/bottom, but numbered - so every value just has a comment with the number pointing to the source, instead of writing the whole thing out every time? Or something else along those lines?

Reply:
Can you paste in a sample of the output as you have it?

Reply:
![ListConfigurationSources_Raw](https://user-images.githubusercontent.com/85592574/152449014-76ec0a4d-8093-4df2-8ac9-932c985473ae.png)
![ListConfigurationSources_References](https://user-images.githubusercontent.com/85592574/152449025-e1146724-64c0-475d-814e-fc2b73259c48.png)

I included screenshots of the two approaches to maintain formatting; this isn't the whole output, but should give you an idea. Currently, the source is appearing after the key/value instead of before it just due to the way we write the JSON; this can probably be worked around, but I expect that would be a bit more intrusive than the current solution.

Reply:
Also, ignore the bug in the first image under "Subject" (I fixed it when I was making the second version); forgot a break statement after finding the provider.

Reply:
Pinging - wanted to make sure people didn't forget about this thread

Reply:
I'm ok with the first one. I don't think the verbosity is really an issue, since you're asking for the information and can use tooling to help analyze the file.

Reply:
Also, please put the comment above the property/value instead under it.